### PR TITLE
[ConstraintSystem] Add Array as a designated type for `+` and `+=`.

### PR DESCRIPTION
**Pull request by rudkx**
Also add overloads for these operators to an extension of Array.

This allows us to typecheck array concatenation quickly with
designated type support enabled and the remaining type checker hacks
disabled.

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
This builds on https://github.com/apple/swift/pull/20618, which is the first commit in this PR.

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Review by xedin — APPROVED** [review]
LGTM!

File: stdlib/public/core/Array.swift [hunk 1]
```diff
@@ -1290,6 +1290,25 @@ extension Array: RangeReplaceableCollection {
   }
 }
 
+// Implementations of + and += for same-type arrays. This combined
+// with the operator declarations for these operators designating this
+// type as a place to prefer this operator help the expression type
+// checker speed up cases where there is a large number of uses of the
+// operator in the same expression.
+extension Array {
+  @inlinable // FIXME(sil-serialize-all)
```
Comment (airspeedswift) [inline stdlib/public/core/Array.swift]: We don't need the FIXMEs here. These are legitimately inlinable.

File: stdlib/public/core/Policy.swift [hunk 2]
```diff
@@ -420,7 +420,7 @@ infix operator   & : MultiplicationPrecedence, BinaryInteger
 
 // "Additive"
 
-infix operator   + : AdditionPrecedence, AdditiveArithmetic, String, Strideable
+infix operator   + : AdditionPrecedence, AdditiveArithmetic, String, Array, Strideable
```
Comment (airspeedswift) [inline stdlib/public/core/Policy.swift]: Unrelated to PR: looks like GitHub syntax highlighting needs updating to handle the new operator syntax.

**Review by airspeedswift — COMMENTED** [review]
LGTM. My only concern is if it breaks the other versions, on range-replaceable, like if it interferes with `var a = [1,2,3]; a += 4...6`. Not sure what testing coverage we have for this usage.

File: stdlib/public/core/Array.swift (hunk 1 above)
Comment (rudkx) [reply stdlib/public/core/Array.swift]: Okay I'll remove those in a separate PR. This was copy/pasted from elsewhere.

**Comment by rudkx** [comment]
It won't interfere with correctness of cases like the one with the range. We'll still fall back on those overloads.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test and merge

**Pull request merged**